sdk/python/feast/cli_utils.py

```diff
@@ -195,7 +198,121 @@ def handle_not_verbose_permissions_command(
             [t.__name__ for t in p.types],  # type: ignore[union-attr, attr-defined]
             p.with_subclasses,
             p.name_pattern,
-            p.actions,
-            roles,
+            [a.value.upper() for a in p.actions],
+            sorted(roles),
         ]
     )
+
+
+def fetch_all_feast_objects(store: FeatureStore) -> list[FeastObject]:
+    objects: list[FeastObject] = []
+    objects.extend(store.list_entities())
+    objects.extend(store.list_all_feature_views())
+    objects.extend(store.list_batch_feature_views())
+    objects.extend(store.list_feature_services())
+    objects.extend(store.list_data_sources())
+    objects.extend(store.list_data_sources())
+    objects.extend(store.list_validation_references())
+    objects.extend(store.list_saved_datasets())
+    objects.extend(store.list_permissions())
+    return objects
+
+
+def handle_permissions_check_command(
+    object: FeastObject, permissions: list[Permission], table: list[Any]
+):
+    for p in permissions:
+        if p.match_resource(object):
+            return
+    table.append(
+        [
+            object.name,
+            type(object).__name__,
+        ]
+    )
+
+
+def handle_permissions_check_command_with_actions(
+    object: FeastObject, permissions: list[Permission], table: list[Any]
+):
+    unmatched_actions = ALL_ACTIONS.copy()
+    for p in permissions:
+        if p.match_resource(object):
+            for action in ALL_ACTIONS:
+                if p.match_actions([action]) and action in unmatched_actions:
+                    unmatched_actions.remove(action)
+
+    if unmatched_actions:
+        table.append(
+            [
+                object.name,
+                type(object).__name__,
+                [a.value.upper() for a in unmatched_actions],
+            ]
+        )
+
+
+def fetch_all_permission_roles(permissions: list[Permission]) -> list[str]:
+    all_roles = set()
+    for p in permissions:
+        if isinstance(p.policy, RoleBasedPolicy) and len(p.policy.get_roles()) > 0:
+            all_roles.update(p.policy.get_roles())
+
+    return sorted(all_roles)
+
+
+def handler_list_all_permissions_roles(permissions: list[Permission], table: list[Any]):
+    all_roles = fetch_all_permission_roles(permissions)
+    for role in all_roles:
+        table.append(
+            [
+                role,
+            ]
+        )
+
+
+def handler_list_all_permissions_roles_verbose(
+    objects: list[FeastObject], permissions: list[Permission], table: list[Any]
+):
+    all_roles = fetch_all_permission_roles(permissions)
+
+    for role in all_roles:
+        for o in objects:
+            permitted_actions = ALL_ACTIONS.copy()
+            for action in ALL_ACTIONS:
+                matching_permissions = [
```

Comment:
This code looks close to the code in the enforcer, can't we reuse it by extracting it from there to a method or a few and reusing it ?

Reply:
yes, it looks similar but it isn't. Let me see if I can make an helper function out of it

Reply:
I tried but the return type is really different and also we do not raise errors in the CLI code. I will add a comment in the code for now